### ISSUE TITLE
Problem reading the set values 

### ISSUE THREAD

**Issue by stcan** [aloishockenschlohe/dpm86_power_supply#3]
Hi,

I have a problem to read the set values.  I can set anything without problems, but not read the values. Here the log for a write operation:

$ ./dpm86 c 15000
./dpm86 version 221009 [2526] - dpm86write -- sending ':01w11=15000.'
./dpm86 version 221009 [2526] - dpm86write -- reading response
./dpm86 version 221009 [2526] - dpm6write -- received response ':01ok'.

And here the log for a reading response:
$ ./dpm86 c 
./dpm86 version 221009 [2576] - dpm86read -- sending: ':01r31=0'
./dpm86 version 221009 [2576] - dpm86read -- reading response
 ./dpm86 version 221009 [2576] - response=5=0. --
./dpm86 version 221009 [2576] - dpm86read -- received response ''

I am using a DPM8624-485 with the included RS485-USB adapter and a Raspberry PI with current "Raspbian GNU/Linux 11 (bullseye)".

Thanks and best
steve

**Comment by aloishockenschlohe** [comment]
Hi!

Thank you for your message.
The power supply seems to respond with the message "5=0.<cut-off-character>" and this is very strange.

In principle, there are three possibilities:

1. The manufacturer has changed the protocol.
2. I have messed up. :-)
3. Raspberry Pi is a scary bad platform with more bugs than features.

Normally I guess 2 but in this case I guess 3.

Why is that?
I made very bad experiences with the Raspberry Pi when programming the tool (or maybe I never understood dtty devices - that is definitely a hint for possibility 2 ). It was really complex until I could reliably exchange data with the power supply (and I don't know until today why some extremely strange errors occurred).

Here is an example:
For quite a while I used the tool 'socat' to send data to the tty device, because it was just faster and more comfortable than using the good old 'echo' command. This worked quite well - for five to seven minutes. After this period of time, the Pi shut down with an error message from the kernel. Reliable. Reproducible. Frustrating.

Long story short:
Please send the output of

* dmesg
* lsusb -v
* lsmod

and we can (maybe) figure out what's going on.

**Comment by stcan** [comment]
Here are the interesting sections of dmesg, lsmod and lsusb -v. 

I tested some more things and had a little more success with the command line arguments "read" and "write". With this it was possible to program values as well as to read them. At least for a certain time, until no more values could be read.

There were also sporadic timeouts, which were not traceable.

In parallel I tested the communication to the DPM8624 with python and it works fine. (Same Hardware)


demsg

```
[   32.895947] usbcore: registered new interface driver usbserial_generic
[   32.896137] usbserial: USB Serial support registered for generic
[   33.058740] usbcore: registered new interface driver ch341
[   33.058944] usbserial: USB Serial support registered for ch341-uart
[   33.059137] ch341 1-1.4:1.0: ch341-uart converter detected
[   33.359173] ch341-uart ttyUSB0: break control not supported, using simulated break
[   33.359737] usb 1-1.4: ch341-uart converter now attached to ttyUSB0
```

lsusb -v

```
Bus 001 Device 005: ID 1a86:7523 QinHeng Electronics CH340 serial converter
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               1.10
  bDeviceClass          255 Vendor Specific Class
  bDeviceSubClass         0
  bDeviceProtocol         0
  bMaxPacketSize0         8
  idVendor           0x1a86 QinHeng Electronics
  idProduct          0x7523 CH340 serial converter
  bcdDevice            2.54
  iManufacturer           0
  iProduct                2 USB2.0-Ser!
  iSerial                 0
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0027
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0
    bmAttributes         0x80
      (Bus Powered)
    MaxPower               96mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           3
      bInterfaceClass       255 Vendor Specific Class
      bInterfaceSubClass      1
      bInterfaceProtocol      2
      iInterface              0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0020  1x 32 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0020  1x 32 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
```




lsmod
```
Module                  Size  Used by
rfcomm                 49152  4
cmac                   16384  2
algif_hash             16384  1
aes_arm                16384  2
aes_generic            40960  1 aes_arm
ecb                    16384  1
algif_skcipher         16384  1
af_alg                 28672  6 algif_hash,algif_skcipher
bnep                   20480  2
snd_soc_hdmi_codec     20480  1
btusb                  45056  0
vc4                   282624  2
btrtl                  20480  1 btusb
btintel                36864  1 btusb
btbcm                  20480  1 btusb
bluetooth             425984  33 btrtl,btintel,bnep,btbcm,rfcomm,btusb
sha256_generic         16384  0
libsha256              16384  1 sha256_generic
cfg80211              782336  0
cec                    49152  1 vc4
8021q                  32768  0
garp                   16384  1 8021q
stp                    16384  1 garp
llc                    16384  2 garp,stp
ch341                  20480  0
usbserial              36864  1 ch341
drm_kms_helper        270336  2 vc4
ecdh_generic           16384  1 bluetooth
ecc                    40960  1 ecdh_generic
rfkill                 32768  4 bluetooth,cfg80211
libaes                 20480  3 bluetooth,aes_arm,aes_generic
snd_soc_core          233472  2 vc4,snd_soc_hdmi_codec
bcm2835_codec          49152  0
v4l2_mem2mem           36864  1 bcm2835_codec
bcm2835_v4l2           45056  0
snd_compress           20480  1 snd_soc_core
bcm2835_isp            28672  0
snd_pcm_dmaengine      16384  1 snd_soc_core
syscopyarea            16384  1 drm_kms_helper
bcm2835_mmal_vchiq     32768  3 bcm2835_isp,bcm2835_codec,bcm2835_v4l2
sysfillrect            16384  1 drm_kms_helper
raspberrypi_hwmon      16384  0
sysimgblt              16384  1 drm_kms_helper
fb_sys_fops            16384  1 drm_kms_helper
videobuf2_vmalloc      16384  1 bcm2835_v4l2
videobuf2_dma_contig    20480  2 bcm2835_isp,bcm2835_codec
videobuf2_memops       16384  2 videobuf2_dma_contig,videobuf2_vmalloc
videobuf2_v4l2         32768  4 bcm2835_isp,bcm2835_codec,bcm2835_v4l2,v4l2_mem2mem
videobuf2_common       61440  8 bcm2835_isp,bcm2835_codec,videobuf2_dma_contig,videobuf2_vmalloc,videobuf2_memops,bcm2835_v4l2,v4l2_mem2mem,videobuf2_v4l2
snd_bcm2835            24576  1
snd_pcm               114688  5 snd_compress,snd_pcm_dmaengine,snd_soc_hdmi_codec,snd_bcm2835,snd_soc_core
videodev              237568  6 bcm2835_isp,bcm2835_codec,videobuf2_common,bcm2835_v4l2,v4l2_mem2mem,videobuf2_v4l2
i2c_bcm2835            16384  0
snd_timer              32768  1 snd_pcm
snd                    81920  8 snd_compress,snd_soc_hdmi_codec,snd_timer,snd_bcm2835,snd_soc_core,snd_pcm
mc                     49152  6 bcm2835_isp,bcm2835_codec,videobuf2_common,videodev,v4l2_mem2mem,videobuf2_v4l2
vc_sm_cma              32768  2 bcm2835_isp,bcm2835_mmal_vchiq
uio_pdrv_genirq        16384  0
uio                    20480  1 uio_pdrv_genirq
fixed                  16384  0
drm                   524288  3 vc4,drm_kms_helper
fuse                  126976  1
drm_panel_orientation_quirks    16384  1 drm
backlight              20480  2 drm_kms_helper,drm
ip_tables              28672  0
x_tables               36864  1 ip_tables
ipv6                  491520  32
```

**Comment by aloishockenschlohe** [comment]
Hi!

I re-wrote dpm86 in python. Please give it a try. :-)

**Comment by stcan** [comment]
Hi,

The Python version runs much better! The programming of voltage and current values is very reliable and there is very rarely an empty value when reading the voltage values. I can live with that. 

The controller script is monitored and restarted if necessary. I plan to check if the measured values are plausible to make the controller script even more robust.

Thanks for the effort!

**Comment by aloishockenschlohe** [comment]
Hi!

First of all I am very glad for your feedback.

After some thinking I finally was hit by some kind of enlightenment.
Since I fixed the two (!) issues I figured out, everything runs like a charm. :-)

Enlightment: It seems to be a simple but very solid physical problem.

**Issue 1**

The problem and the fix packed together in one sentence:
Please make sure that there is no potential difference between the power supply and the controlling device.

Otherwise the signal transmission is damn error-prone, which becomes noticeable e.g. by strange characters in the transmission (or no transmission at all). 

How do you make sure of that?
Better USB-to-RS485 adapters have their own connector ("GND") for this purpose, which you can connect to the corresponding counterpart on the power supply.

Cheap adapters (€ 4,95) do not have this port. If you have a cheap device, then it may help to connect the "GND" connector of the power supply to the ground of the power connector where the power supply of the controlling device is plugged in. 

How can you test if you a struggling with the described problem?
Open a shell and type:

> pi@raspberry:~ $ while true; do ./dpm86ng voltage; done

If you see errors: Bingo!

**Issue 2**
Because I can now read out the parameters really, really, really fast I ran into another issue: The terminal characters of the commands were kind of wrong. I fixed this problem in the latest dpm86_python. Please update!!!

**Comment by aloishockenschlohe** [comment]
Okay. Another problem showed up:

The kernel module "ch341" included in the mainstream kernel is rather buggy. After a while a error message shows up ("serial.serialutil.SerialException: [Errno 12] could not open port /dev/ttyUSB0: [Errno 12] Cannot allocate memory: '/dev/ttyUS0'") and the serial port isn't reachable anymore. At the moment an alernative kernel module is tested (see https://github.com/WCHSoftGroup/ch341ser_linux). Press your thumbs!

**Comment by stcan** [comment]
Hi, I updated your script and typed:
`pi@raspberry:~ $ while true; do ./dpm86 voltage; done`

The loop ran through several hundred times without error.  

And I have exactly the cheap 2 pole adapter, only "A" and "B", no GND connection. It is a feature of the RS485 to get along without GND and in my case it works very well.  I use an unshielded twisted pair cable, which is not susceptible to interference. Maybe that's why it works so well.

I will test the controller script with your latest dpm86 version and report back.

**Comment by aloishockenschlohe** [comment]
I checked various other things meanwhile.
And I think I found the root cause: Raspberry Pi 4.

After I switched back to RPi 3B, everything works stable.
I close this issue.
